### DISCUSSION TITLE
Info Request: Nonlinear Processes using the adjoint solver

### DISCUSSION THREAD

**Discussion by Andeloth** [NanoComp/meep#2094] [Q&A] [answered]
I'd like to use meep's adjoint solver to optimize a nonlinear process. Does the adjoint solver's current implementation support using nonlinear materials in the design region?

I've tried to run an optimization where one of the materials in the design region had nonlinearities, but the gradient calculation seemed to stall for several hours, I assumed because it was taking too long to take the gradients. My figure of merit function was a simple surface flux at a different wavelength.

Do you know if this was caused by nonlinearities not being implemented in the adjoint solver's gradient calculations, or is something else causing it?

**Comment by smartalecH** [comment, marked as answer]
Gradients for nonlinear problems are not currently supported (out of the box). That would require a different adjoint-variable formulation. The problem is further complicated by the fact we implement a hybrid time-/frequency-domain adjoint formulation (although some nonlinear processes, like second harmonic generation, could still be modeled with the existing framework eg [using coupled linear solves](https://doi.org/10.1364/OPTICA.3.000233)). 

[Our recent paper](https://doi.org/10.1364/OE.442074) discusses these nuances in detail (see the conclusion).
